XmlFile.js

```diff
@@ -0,0 +1,915 @@
+    var unescaped = XmlFile.unescapeString(string);
+
+    unescaped = unescaped.
+        replace(/[ \n\t\r\f]+/g, " ").
+        trim();
+
+    return unescaped;
+};
+
+
+XmlFile.escapeProp = function(prop) {
+    return prop.
+        replace(/~/g, "~0").
+        replace(/\//g, "~1");
+};
+
+XmlFile.unescapeProp = function(prop) {
+    return prop.
+        replace(/~1/g, "/").
+        replace(/~0/g, "~");
+};
+
+XmlFile.escapeRef = function(prop) {
+    return XmlFile.escapeProp(prop).
+        replace(/%/g, "%25").
+        replace(/\^/g, "%5E").
+        replace(/\|/g, "%7C").
+        replace(/\\/g, "%5C").
+        replace(/"/g, "%22").
+        replace(/ /g, "%20");
+};
+
+XmlFile.unescapeRef = function(prop) {
+    return XmlFile.unescapeProp(prop.
+        replace(/%5E/g, "^").
+        replace(/%7C/g, "|").
+        replace(/%5C/g, "\\").
+        replace(/%22/g, "\"").
+        replace(/%20/g, " ").
+        replace(/%25/g, "%"));
+};
+
+function isPrimitive(type) {
+    return ["boolean", "number", "integer", "string"].indexOf(type) > -1;
+}
+
+/**
+ * Gets type of array based on provided schema.
+ *
+ * TODO: Add support for "anyOf" and "oneOf" type definitions.
+ */
+function getArrayTypeFromSchema(schema, root) {
+    if (schema.type !== "array") {
+        return null;
+    }
+
+    if (typeof(schema.items["$ref"]) !== 'undefined') {
+        // substitute the referenced schema for this one
+        var refname = schema.items["$ref"];
+        var otherschema = root["$$refs"][refname];
+        if (!otherschema) {
+            console.log("Unknown reference " + refname + " while parsing " +
+                    this.pathName + " with schema " + root["$id"]);
+            return;
+        }
+        schema = otherschema;
+    } else {
+        schema = schema.items;
+    }
+
+    var allowedTypes = ["string", "integer", "number", "boolean", "object"];
+
+    if (schema.type && allowedTypes.indexOf(schema.type) > -1) {
+        return schema.type;
+    }
+
+    // Default type is string for compatibility reasons.
+    return 'string';
+}
+
+/**
+ * Converts each element of one-dimensional array to a primitive type.
+ */
+function convertArrayElementsToType(array, type) {
+    if (!ilib.isArray(array)){
+        return array;
+    }
+
+    return array.map(function (item) {
+        return convertValueToType(item, type);
+    });
+}
+
+/**
+ * Converts value to a primitive type.
+ */
+function convertValueToType(value, type) {
+    switch (type) {
+        case "boolean":
+            return value === "true";
+        case "number":
+            return Number.parseFloat(value);
+        case "integer":
+            return Number.parseInt(value);
+        default:
+            return value;
+    }
+}
+
+var pluralCategories = {
+    "zero": true,
+    "one": true,
+    "two": true,
+    "few": true,
+    "many": true,
+    "other": true
+};
+
+function isNotEmpty(obj) {
+    if (!obj) {
+        return false;
```

Comment:
This place is subjected to a bug that was fixed there: https://github.com/iLib-js/ilib-loctool-json/commit/4d6b318ddd0c20f6e7442cb5e529c57a91f004ed

Object that contains only boolean values is completely ignored if all values equal `false`.

Example:
```
{
    "nullable": false
}
```
Will disappear from the output since this object is considered "empty". Please evaluate how this might affect the XML plugin and decide if that needs to be fixed.

Reply:
Fix copied from ilib-loctool-json